src/main/java/com/wavefront/sdk/entities/histograms/WavefrontHistogramImpl.java

```diff
@@ -377,20 +308,66 @@ public Distribution(long timestamp, List<Pair<Double, Integer>> centroids) {
   /**
    * Representation of a bin that holds histogram data for a particular minute in time.
    */
-  private static class MinuteBin {
+  private static class ThreadMinuteBin {
     /**
-     * The histogram data for the minute bin, represented as a {@link TDigest} distribution
+     * Stores the {@link TDigest} Distribution for each threads in one given minute.
      */
-    public final TDigest distribution;
+    public ConcurrentMap<Long, TDigest> perThreadDist;
 
     /**
      * The timestamp at the start of the minute.
      */
     public final long minuteMillis;
 
-    MinuteBin(int accuracy, long minuteMillis) {
-      distribution = new AVLTreeDigest(accuracy);
+    ThreadMinuteBin(long minuteMillis) {
+      perThreadDist = new ConcurrentHashMap<>();
       this.minuteMillis = minuteMillis;
     }
+
+    /**
+     * Copy Constructor for appending ThreadMinuteBin to globalHistogramBinsList
+     */
+    ThreadMinuteBin(ThreadMinuteBin threadMinuteBin) {
+      this.perThreadDist = new ConcurrentHashMap<>(threadMinuteBin.perThreadDist);
+      this.minuteMillis = threadMinuteBin.minuteMillis;
+    }
+
+    /**
+     * Helper to retrieve the thread-local distribution in one given minute.
+     */
+    TDigest getDistByThreadId(long threadId) {
+      // Create new Digest for new thread.
+      return perThreadDist.computeIfAbsent(threadId, digest -> new AVLTreeDigest(ACCURACY));
```

Comment:
should be threadId, key -> new AVLTreeDigest(ACCURACY)
as digest is not long, 
timestampMillis is the key here that is of type long.